File: src/main/java/org/modelingvalue/nelumbo/And.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class And extends BinaryPredicate {
    @Serial
    private static final long serialVersionUID = -7248491569810098948L;

    private static Functor    FUNCTOR;

    static {
        KnowledgeBase.registerFunctorSetter(And.class, f -> FUNCTOR = f);
    }

    public And(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args[0], args[1]);
    }

    private And(Object[] args, And declaration) {
        super(args, declaration);
    }

    public static And of(Node predicate1, Node predicate2) {
        return new And(FUNCTOR, List.of(), new Object[]{predicate1, predicate2});
    }

    @Override
    public And declaration() {
        return (And) super.declaration();
    }

    @Override
    protected And struct(Object[] array, Predicate declaration) {
        return new And(array, (And) declaration);
    }

    @Override
    public And set(int i, Object... a) {
        return (And) super.set(i, a);
    }

    @Override
    protected boolean isTrue(InferResult predResult) {
        return false;
    }

    @Override
    protected boolean isFalse(InferResult predResult) {
        return predResult.isFalseCC();
    }

    @Override
    protected boolean isTrue(InferResult[] predResult) {
        return predResult[0].isTrueCC() && predResult[1].isTrueCC();
    }

    @Override
    protected boolean isFalse(InferResult[] predResult) {
        return false;
    }

    @Override
    protected boolean isLeft(InferResult[] predResult) {
        return predResult[1].isTrueCC();
    }

    @Override
    protected boolean isRight(InferResult[] predResult) {
        return predResult[0].isTrueCC();
    }

    @Override
    protected InferResult add(InferResult[] predResult) {
        return predResult[0].addAnd(predResult[1]);
    }

}

File: web-ts/src/logic/And.ts
/**
 * And - logical AND predicate.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.And
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { BinaryPredicate } from './BinaryPredicate';
import { Predicate } from './Predicate';
import { InferResult } from '../InferResult';

/**
 * And - logical conjunction.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.And
 */
export class And extends BinaryPredicate {
  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#FUNCTOR
  static FUNCTOR: Functor | null = null;

  constructor(functor: Functor, elements: List<AstElement>, left: Node, right: Node) {
    super(functor, elements, left, right);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#of(Predicate, Predicate)
  static of(predicate1: Predicate, predicate2: Predicate): And {
    return new And(And.FUNCTOR!, List<AstElement>(), predicate1, predicate2);
  }

  protected static fromDataAnd(data: unknown[], declaration?: Node): And {
    const and = Object.create(And.prototype) as And;
    (and as unknown as { _data: unknown[] })._data = data;
    (and as unknown as { _declaration: Node })._declaration = declaration ?? and;
    (and as any)._binding = null;
    (and as any)._hashCodeCached = false;
    (and as any)._hashCode = 0;
    (and as any)._nrOfUnbound = -1;
    return and;
  }

  protected override struct(data: unknown[], declaration?: Node): And {
    return And.fromDataAnd(data, declaration ?? this.declaration());
  }

  override declaration(): And {
    return super.declaration() as And;
  }

  override set(i: number, ...a: unknown[]): And {
    return super.set(i, ...a) as And;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isTrue(InferResult, int)
  protected isTrueSingle(_predResult: InferResult, _i: number): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isFalse(InferResult, int)
  protected isFalseSingle(predResult: InferResult, _i: number): boolean {
    return predResult.isFalseCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isUnknown(InferResult, int)
  protected isUnknownSingle(_predResult: InferResult, _i: number): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isTrue(InferResult[])
  protected isTrueBoth(predResult: InferResult[]): boolean {
    return predResult[0].isTrueCC() && predResult[1].isTrueCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isFalse(InferResult[])
  protected isFalseBoth(_predResult: InferResult[]): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isLeft(InferResult[])
  protected isLeft(predResult: InferResult[]): boolean {
    return predResult[1].isTrueCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.And#isRight(InferResult[])
  protected isRight(predResult: InferResult[]): boolean {
    return predResult[0].isTrueCC();
  }
}
